Redesign of the DatePartitionedQueryPlanner (formerly know as the FederatedQueryPlanner)

ivakegg opened the pull request.

hlgp committed:
investigating why index holes are not being applied

ivakegg committed:
Changed IndexHole to ValueIndexHole
Updated TODOs

ivakegg committed:
Changed getters and setters from indexHoles to valueIndexHoles

ivakegg committed:
Changed FederatedQueryPlanner to DatePartitionedQueryPlanner

ivakegg committed:
More renaming of stuff to make Hannah happy

ivakegg committed:
Working through new algorithm

ivakegg committed:
Redesigned process by which index holes are determined such that we know
what fields are unindexed in each range

ivakegg committed:
Merge branch 'integration' into task/hp-indexHoleInvestigation

ivakegg committed:
merge conflict updates

ivakegg committed:
Added reprocessing of a query plan for subplans

ivakegg committed:
Updated to handle expansion after pullup

ivakegg committed:
7.16.0-planner

ivakegg committed:
updated javadoc

ivakegg committed:
Merge branch 'integration' into task/hp-indexHoleInvestigation

ivakegg committed:
Revert "7.16.0-planner"

This reverts commit 549130ba3c3aef26f0bab1b3e25ca622e09510f3.

hlgp commented on warehouse/query-core/src/main/java/datawave/query/planner/DatePartitionedQueryPlanner.java:
```diff
@@ -0,0 +1,820 @@
+
+    /**
+     * Return the set of date ranges that sub-queries should be created for. Each date range will have a consistent index state, meaning that within each date
+     * range there will be zero or more fields that are completely indexed in that range. Also it is expected that the date ranges will complete cover the query
```
nit: completely cover

ivakegg replied:
reworded

hlgp commented on warehouse/query-core/src/main/java/datawave/query/planner/DatePartitionedQueryPlanner.java:
```diff
@@ -0,0 +1,820 @@
+
+    /**
+     * This method is intended to ensure some fault tolerance in out production of the timeline. The date range from beginDate to endDate should be completely
+     * covered, there should be no gaps, no overlapping date ranges, no negative length date ranges, and every date range should have a different set of
```
yes!  love the validation

ivakegg replied:
thank you

hlgp commented on warehouse/query-core/src/test/java/datawave/query/IndexHoleQueryTest.java:
```diff
@@ -15,7 +15,7 @@
 
-import datawave.query.config.IndexHole;
+import datawave.query.config.IndexValueHole;
 import datawave.query.exceptions.FullTableScansDisallowedException;
```
Rename test class to IndexValueHoleQueryTest?

ivakegg replied:
done

ivakegg committed:
A little refactoring

ivakegg committed:
Added some javadoc

hlgp commented:
We should probably adjust the marker concept to be one day earlier than the last old index entry timestamp.  Since we usually deploy during the day, any deployment where we ceased indexing something would only be a partial day.  I replicated this with a test case and any data from the last indexed date that did not get indexed (i.e. post hypothetical deployment) would be missed.

hlgp commented:
In other tests, I'm getting empty ranges.  We're still checking the fi for the field with the _Hole_ marker, so the AndIterator appears to be defeating.  Haven't fully stepped through, though, but we need more end-to-end tests up in here (up in here).

apmoriarty commented on warehouse/query-core/src/main/java/datawave/query/jexl/visitors/PushdownUnindexedFieldsVisitor.java:
```diff
@@ -0,0 +1,171 @@
+import datawave.query.util.MetadataHelper;
+
+/**
+ * Visitor meant to 'push down' predicates for expressions that are not executable against the index.
```
index!? global index or field index.

ivakegg replied:
updated comment.

apmoriarty commented on warehouse/query-core/src/main/java/datawave/query/jexl/visitors/PushdownUnindexedFieldsVisitor.java:
```diff
@@ -0,0 +1,171 @@
+/**
+ * Visitor meant to 'push down' predicates for expressions that are not executable against the index.
+ */
+public class PushdownUnindexedFieldsVisitor extends RebuildingVisitor {
```
RebuildingVisitor isn't necessary. If you have 100 terms in a query and only one unindexed term, you're copying 100 terms to swap in a single new term. Extend base visitor and modify the query tree in place, with confidence!

ivakegg replied:
I have to use the rebuilding visitor.  Each set of unindexed fields is being applied to the original query plan tree so I cannot modify the tree inline.

apmoriarty commented on warehouse/query-core/src/main/java/datawave/query/jexl/visitors/PushdownUnindexedFieldsVisitor.java:
```diff
@@ -0,0 +1,171 @@
+        return node;
+    }
+
+    public boolean isIndexed(JexlNode node) {
+        String field = JexlASTHelper.getIdentifier(node);
+        try {
+            return (field != null && this.helper.isIndexed(field, this.dataTypeFilter));
+        } catch (TableNotFoundException e) {
+            throw new IllegalStateException("Unable to find metadata table", e);
+        }
+    }
```
I question whether the metadata helper is actually necessary for this. You already have a set of non-indexed fields...by definition isn't every field not in that set indexed? If you do need to determine the indexed state of a field, why not remove the metadata helper and simply add the set of indexed fields? Then you can remove the set of non-indexed fields.

ivakegg replied:
I will see what I can do here.

ivakegg replied:
Actually there is no need to check for indexed fields as the unindexed field list is enough for this visitor..... fixing.

apmoriarty commented on warehouse/query-core/src/main/java/datawave/query/jexl/visitors/PushdownUnindexedFieldsVisitor.java:
```diff
@@ -0,0 +1,171 @@
+    @Override
+    public Object visit(ASTReferenceExpression node, Object data) {
+        // if not already delayed somehow
+        if (!QueryPropertyMarker.findInstance(node).isAnyTypeExcept(BOUNDED_RANGE)) {
+            return super.visit(node, data);
+        }
```
Marker nodes are handled at the level of an ASTAndNode, you are not guaranteed to have a reference expression. You already handle this case above, so this method isn't necessary

ivakegg replied:
you are probably correct.  I will validate.

apmoriarty commented on warehouse/query-core/src/main/java/datawave/query/jexl/visitors/PushdownUnindexedFieldsVisitor.java:
```diff
@@ -0,0 +1,171 @@
+    }
+
+    @Override
+    public Object visit(ASTERNode node, Object data) {
+        if (isIndexed(node) && missingIndexRange(node)) {
```
extract the field once and pass it to the helper methods. helper methods should operate on the field.

ivakegg replied:
ok, fair point.  thank you!

apmoriarty commented on warehouse/query-core/src/main/java/datawave/query/jexl/visitors/PushdownUnindexedFieldsVisitor.java:
```diff
@@ -0,0 +1,171 @@
+
+    @Override
+    public Object visit(ASTEQNode node, Object data) {
+        if (isIndexed(node) && missingIndexRange(node)) {
```
I'm not sure that this logic will handle queries with arithmetic or methods very well. The field or the value may be null.
See the FunctionalSetTest for examples

ivakegg replied:
but those methods will not be used in the index lookup so do they matter?

ivakegg replied:
Sorry, I think, I see what you mean.... I will reevaluate what I am missing.

ivakegg replied:
yeah, totally rewritting this one.  I had copied it from the other pushdown visitor we had and did not think hard enough about it.  Also writting some tests.

apmoriarty commented on warehouse/query-core/src/main/java/datawave/query/jexl/visitors/PushdownUnindexedFieldsVisitor.java:
```diff
@@ -0,0 +1,171 @@
+
+    private static final Logger log = Logger.getLogger(PushdownUnindexedFieldsVisitor.class);
+
+    protected ShardQueryConfiguration config;
```
config isn't necessary

ivakegg replied:
cool

ivakegg committed:
Changed from using _HOLE_ to _EVAL_ to avoid field index lookups
review comments

ivakegg commented:
> In other tests, I'm getting empty ranges. We're still checking the fi for the field with the _Hole_ marker, so the AndIterator appears to be defeating. Haven't fully stepped through, though, but we need more end-to-end tests up in here (up in here).

I changed this to use evaluation only instead of index value hole markers in the query to avoid the field index lookups for unindexed fields.

ivakegg committed:
Simplified unindexed fields visitor per review comments

lbschanno commented on warehouse/query-core/src/main/java/datawave/query/planner/DatePartitionedQueryPlanner.java:
```diff
@@ -0,0 +1,801 @@
+
+    /**
+     * Return the set of date ranges that sub-queries should be created for. Each date range will have a consistent index state, meaning that within each date
+     * range all query fields are either indexed or not-indexed across the entire range. It is expected that the date ranges will complete cover the original
```
I think you intended `completely cover`, not `complete cover`.

lbschanno commented on warehouse/query-core/src/main/java/datawave/query/planner/DatePartitionedQueryPlanner.java:
```diff
@@ -0,0 +1,801 @@
+    private final Set<String> plans = new LinkedHashSet<>();
+    private DefaultQueryPlanner queryPlanner;
+    private String plannedScript;
+    private SortedSet<Pair<Date,Date>> relevantHoles;
```
As far as I can tell, this isn't being used anywhere.

lbschanno commented on warehouse/query-core/src/main/java/datawave/query/planner/DatePartitionedQueryPlanner.java:
```diff
@@ -0,0 +1,801 @@
+    private SortedSet<Pair<Date,Date>> relevantHoles;
+
+    // handles boilerplate operations that surround a visitor's execution (e.g., timers, logging, validating)
+    private TimedVisitorManager visitorManager = new TimedVisitorManager();
```
This can be final.

lbschanno commented on warehouse/query-core/src/main/java/datawave/query/planner/FederatedQueryIterable.java:
```diff
@@ -9,7 +9,7 @@
- * Implementation of {@link CloseableIterable} intended to be used by {@link FederatedQueryPlanner}. This iterable
+ * Implementation of {@link CloseableIterable} intended to be used by {@link DatePartitionedQueryPlanner}. This iterable
  */
 public class FederatedQueryIterable implements CloseableIterable<QueryData> {
```
Should this be renamed to `DatePartitionedQueryIterable`?

lbschanno commented on warehouse/query-core/src/main/java/datawave/query/planner/DatePartitionedQueryPlanner.java:
```diff
@@ -0,0 +1,801 @@
+        } else if (this.plans.size() == 1) {
+            this.plannedScript = this.plans.iterator().next();
+        } else {
+            int lastIndex = plans.size() - 1;
```
`lastIndex` is never used and can be deleted.

lbschanno commented on warehouse/query-core/src/main/java/datawave/query/planner/DatePartitionedQueryPlanner.java:
```diff
@@ -0,0 +1,801 @@
+        public int compareTo(IndexFieldHoleBoundary other) {
+            int comparison = date.compareTo(other.date);
+            if (comparison == 0) {
+                comparison = Boolean.valueOf(other.indexFieldHoleStart).compareTo(Boolean.valueOf(indexFieldHoleStart));
```
This can be simplified to `Boolean.compare(boolean, boolean)`.

lbschanno commented on warehouse/query-core/src/main/java/datawave/query/planner/DatePartitionedQueryPlanner.java:
```diff
@@ -0,0 +1,801 @@
+     * @param endDate
+     *            The end date
+     */
+    private void ensureConsistency(SortedMap<Pair<Date,Date>,Set<String>> timeline, Date beginDate, Date endDate) throws DatawaveFatalQueryException {
```
I really like this.

lbschanno commented on warehouse/query-core/src/main/java/datawave/query/planner/DatePartitionedQueryPlanner.java:
```diff
@@ -0,0 +1,801 @@
+            String field = hole.getKey();
+            IndexFieldHole indexHole = hole.getValue();
+            for (Pair<Date,Date> range : getHolesOverlappingOriginalQueryDateRange(beginDate, endDate, indexHole)) {
+                timeline.add(new IndexFieldHoleBoundary(range.getLeft(), true, field));
```
`IndexFieldHoleBoundary` can hold a set of fields, however only the single field constructor `IndexFieldHoleBoundary(Date date, boolean start, String field)` is used here, and `IndexFieldHoleBoundary.addFields()` is never called. 

Did you intend for `createTimeline()` to return a set of boundaries that contain singular fields only, or are the fields supposed to be collapsed into any boundaries where they overlap?

ivakegg replied:
You are right, I was thinking of collapsing the boundaries but I ended up not needing/wanting to do that.....  removing obsolete functionality.

lbschanno commented on warehouse/query-core/src/main/java/datawave/query/planner/DatePartitionedQueryPlanner.java:
```diff
@@ -0,0 +1,801 @@
+            timeline.add(new IndexFieldHoleBoundary(endDate, false));
+        } else {
+            if (timeline.first().getBoundary().after(beginDate)) {
+                // start with and beginning boundary that starts at the beginDate with no unindex field
```
Minor nitpick: `start with a beginning`, `with no unindexed field`

lbschanno commented on warehouse/query-core/src/main/java/datawave/query/planner/DatePartitionedQueryPlanner.java:
```diff
@@ -0,0 +1,801 @@
+        return new Date(date.getTime() - 1);
+    }
+
+    public static class IndexFieldHoleBoundary implements Comparable<IndexFieldHoleBoundary> {
```
Can you add a comment here describing that this represents a bounding endpoint for the dates that a field index hole encompasses? It becomes apparent after some inspection, but a description never hurts.

lbschanno commented on warehouse/query-core/src/main/java/datawave/query/planner/DatePartitionedQueryPlanner.java:
```diff
@@ -0,0 +1,801 @@
+
+    public static class IndexFieldHoleBoundary implements Comparable<IndexFieldHoleBoundary> {
+        private final Date date;
+        private final boolean indexFieldHoleStart;
```
I don't think this needs to include `indexFieldHole` in the name, `isStart` or `start` would be fine, or perhaps `boundaryStart`.

lbschanno commented on warehouse/query-core/src/main/java/datawave/query/jexl/visitors/PushdownUnindexedFieldsVisitor.java:
```diff
@@ -0,0 +1,131 @@
+     *            the current node
+     * @return a jexl node
+     */
+    protected JexlNode delayBoundedIndexHole(LiteralRange range, ASTAndNode currentNode, Object data) {
```
The parameter `data` is not used and can be removed.

lbschanno commented on warehouse/query-core/src/main/java/datawave/query/jexl/visitors/PushdownUnindexedFieldsVisitor.java:
```diff
@@ -0,0 +1,131 @@
+ */
+public class PushdownUnindexedFieldsVisitor extends RebuildingVisitor {
+
+    private static final Logger log = Logger.getLogger(PushdownUnindexedFieldsVisitor.class);
```
`log` is never used and can be deleted.

lbschanno commented on warehouse/query-core/src/main/java/datawave/query/jexl/visitors/PushdownUnindexedFieldsVisitor.java:
```diff
@@ -0,0 +1,131 @@
+    public Object visit(ASTAndNode node, Object data) {
+        // if not already delayed somehow
+        if (!QueryPropertyMarker.findInstance(node).isAnyTypeExcept(BOUNDED_RANGE)) {
+            LiteralRange range = JexlASTHelper.findRange().getRange(node);
```
This can be simplified to `if(QueryPropertyMarker.findInstance(node).isType(BOUNDED_RANGE)) {`

lbschanno commented on warehouse/query-core/src/main/java/datawave/query/jexl/visitors/PushdownUnindexedFieldsVisitor.java:
```diff
@@ -0,0 +1,131 @@
+/**
+ * Visitor meant to 'push down' predicates for expressions that are not executable against the global OR field index.
+ */
+public class PushdownUnindexedFieldsVisitor extends RebuildingVisitor {
```
Will you be adding any tests for this visitor?

ivakegg replied:
TBD

ivakegg replied:
Yes, will do

lbschanno commented on warehouse/query-core/src/main/java/datawave/query/planner/DatePartitionedQueryPlanner.java:
```diff
@@ -0,0 +1,801 @@
+                                        + dateRange.getValue() + ": " + configCopy.getQueryString());
+                    }
+                } catch (DatawaveQueryException e) {
+                    log.warn("Exception occured when processing sub-plan against date range (" + subBeginDate + "-" + subEndDate + ")", e);
```
Spelling: `occured` should be `occurred`.

ivakegg committed:
Updated per review comments

ivakegg committed:
Added additional logic to the unindexed pushdown visitor and added test
cases.
Refactored part of the JexlStringBuildingVisitor to simplify adding some
missing operators.

ivakegg committed:
Missed a copy

ivakegg committed:
Use actual metadata utils version

apmoriarty commented on warehouse/query-core/src/main/java/datawave/query/jexl/visitors/PushdownUnindexedFieldsVisitor.java:
```diff
@@ -0,0 +1,226 @@
+
+    @Override
+    public Object visit(ASTMethodNode node, Object data) {
+        // no need to delay functions
```
nit pick: methods

apmoriarty commented on warehouse/query-core/src/main/java/datawave/query/jexl/visitors/PushdownUnindexedFieldsVisitor.java:
```diff
@@ -0,0 +1,226 @@
+        return (unindexedFields.contains(range.getFieldName()));
+    }
+
+}
```
This looks much cleaner

apmoriarty commented on warehouse/query-core/src/main/java/datawave/query/jexl/visitors/JexlStringBuildingVisitor.java:
```diff
@@ -188,156 +191,54 @@ public Object visit(ASTAndNode node, Object data) {
-        sb.append(" != ");
-
-        node.jjtGetChild(1).jjtAccept(this, sb);
+    public Object visit(ASTEQNode node, Object data) {
```
This class has some great cleanup

apmoriarty commented on warehouse/query-core/src/test/java/datawave/query/CompositeFunctionsTest.java:
```diff
@@ -670,6 +671,28 @@ public void testTLDWithLuceneAndIdentifierToLiteralEQJexl() throws Exception {
+        eventQueryLogic.setQueryPlanner(new DatePartitionedQueryPlanner());
+        Map<String,String> extraParameters = new HashMap<>();
+        extraParameters.put("include.grouping.context", "true");
+        //TODO: why is HOLE not being pushed down with _Hole_ marker?!
```
Is this still an issue?

hlgp replied:
nope.  corresponding MR: https://github.com/NationalSecurityAgency/datawave/pull/2759

I know you like MRs, so I put some MRs in your MRs.

apmoriarty commented on warehouse/query-core/src/test/java/datawave/query/planner/DatePartitionedQueryPlannerTest.java:
```diff
@@ -638,7 +635,6 @@ public void testFieldIndexMinThresholdWithSomeNotMeeting() throws Exception {
 
-        assertQueryResults();
-        assertSubrangesCorrect();
+        assertSubrangesCorrect(assertQueryResults());
```
It took a bit of reading to understand how this test was handling asserting the correct. I would like to see the final query plans asserted, just so it's easy to visually inspect the final results. Something like
```
expectPlan("20130101", "UUID == 'uuid'")
expectPlan("20130101", "(_Hole_ = true) && (UUID == 'uuid')")
```
Doesn't have to be done before a merge, but should be done relatively soon.

ivakegg replied:
done

ivakegg committed:
No need for deprecated method

ivakegg committed:
Merge remote-tracking branch 'origin/integration' into task/hp-indexHoleInvestigation

ivakegg committed:
Upgrade datawave-metadata-utils to 4.0.9 with GroupingTest fix (#2767)

hlgp committed:
add query test for indexed field hole (#2759)

Co-authored-by: hlgp <hlgp@users.noreply.github.com>

ivakegg committed:
Merge branch 'integration' into task/hp-indexHoleInvestigation

ivakegg committed:
nit

ivakegg committed:
Merge branch 'task/hp-indexHoleInvestigation' of github.com:NationalSecurityAgency/datawave into task/hp-indexHoleInvestigation

ivakegg committed:
unused imports

ivakegg committed:
Updated to latest metadata-utils submodule

ivakegg committed:
Added deprecated class, and potential ingest for better test coverage

ivakegg committed:
Added tests for plans in the date partitioned query planner test

ivakegg committed:
Merge remote-tracking branch 'origin/integration' into task/hp-indexHoleInvestigation

ivakegg committed:
Unused tests

ivakegg committed:
unused imports

ivakegg committed:
refactor some of the ingest to allow dynamically reducing metadeata and
index entries

ivakegg committed:
Refactored the reverse index entries

Commit:
GitHub Actions: Fix Formatting

ivakegg committed:
Removed excess mutations

ivakegg committed:
Merge branch 'task/hp-indexHoleInvestigation' of github.com:NationalSecurityAgency/datawave into task/hp-indexHoleInvestigation

ivakegg committed:
Updated to drop field index entries as well

ivakegg committed:
Enabled dropping field index entries and subsequent test updates
Enabled testing with and without full table scans enabled

ivakegg committed:
Added test for initial plan

ivakegg committed:
Reduce debugging

ivakegg committed:
Merge remote-tracking branch 'origin/integration' into task/hp-indexHoleInvestigation

ivakegg committed:
Removed metadata-utils repo based on release/version4.0.9

ivakegg committed:
git subrepo clone git@github.com:NationalSecurityAgency/datawave-metadata-utils.git core/utils/metadata-utils

subrepo:
  subdir:   "core/utils/metadata-utils"
  merged:   "08198923b4"
upstream:
  origin:   "git@github.com:NationalSecurityAgency/datawave-metadata-utils.git"
  branch:   "main"
  commit:   "08198923b4"
git-subrepo:
  version:  "0.4.9"
  origin:   "https://github.com/ingydotnet/git-subrepo"
  commit:   "cce3d93"

ivakegg committed:
Merge remote-tracking branch 'origin/integration' into task/hp-indexHoleInvestigation

hlgp commented on warehouse/query-core/src/test/java/datawave/query/util/FieldIndexHoleDataIngest.java:
IndexFieldHoleDataIngest

ivakegg replied:
done

ivakegg committed:
Updated ingest name to be consistent

ivakegg committed:
Merge remote-tracking branch 'origin/integration' into task/hp-indexHoleInvestigation
